### PR TITLE
DATA: Corrected GenericRepository Class

### PR DESCRIPTION
**Pull request by Dilshod-Abdijabbarov**
DATA: Corrected GenericRepository Class

### CODE REVIEW

**Comment by Dilshod-Abdijabbarov** [comment]
DATA: Corrected GenericRepository Class

**Review by MuhammadKasimov — CHANGES_REQUESTED** [review]
Make new method which saves changes

**Comment by MuhammadKasimov** [comment]
Create Issue!

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    {
+        private readonly BelissimoDbContext _dbContext;     
+        public GenericRepository(BelissimoDbContext dbContext)
+        {
+            _dbContext = dbContext;
+        }
+        public async ValueTask<T> CreateAsync(T entity)
+        {
+            return (await _dbContext.Set<T>().AddAsync(entity)).Entity;
+            _dbContext.SaveChangesAsync();
```
[the remark this answers is not in the document]
Comment (Dilshod-Abdijabbarov) [reply BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:25]: Whether to add the await keyword to the code

[the remark this answers is not in the document]
Comment (MuhammadKasimov) [reply BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:25]: noo

[the remark this answers is not in the document]
Comment (MuhammadKasimov) [reply BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:25]: create new method to save changes reomve save changes from Create Update Delete

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 2]
```diff
@@ -22,7 +22,7 @@ public GenericRepository(BelissimoDbContext dbContext)
         public async ValueTask<T> CreateAsync(T entity)
         {
             return (await _dbContext.Set<T>().AddAsync(entity)).Entity;
-            _dbContext.SaveChangesAsync();
+            SaveChangesAsync();
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:25]: we will use it in service layer just remove this line

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 3]
```diff
@@ -35,7 +35,7 @@ public async ValueTask<bool> DeleteAsync(Expression<Func<T, bool>> expression)
             else
             {
                 _dbContext.Set<T>().Remove(entity);
-                await _dbContext.SaveChangesAsync();
+                SaveChangesAsync();
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:38]: the same
remove here

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 4]
```diff
@@ -53,7 +53,11 @@ public async ValueTask<T> GetAsync(Expression<Func<T, bool>> expression, string[
         public T Update(T entity)
         {
             return (_dbContext.Set<T>().Update(entity)).Entity;
-            _dbContext.SaveChanges(); 
+            SaveChangesAsync();
+        }
+        public async Task SaveChangesAsync()
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:58]: make it return ValueTask instead of Task

**Review by MuhammadKasimov — CHANGES_REQUESTED** [review]

**Comment by MuhammadKasimov** [comment]
Issue!

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 5]
```diff
@@ -14,50 +14,47 @@ namespace BelissimoCloneWPF.Data.Repositories
...
 
         public async ValueTask<T> GetAsync(Expression<Func<T, bool>> expression, string[] includes = null)
         {
-            return await _dbContext.Set<T>().FirstOrDefaultAsync(expression);
+            return await dbContext.Set<T>().FirstOrDefaultAsync(expression);
         }
 
         public T Update(T entity)
         {
-            return (_dbContext.Set<T>().Update(entity)).Entity;
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs]: Crete DbSet field instead

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 6]
```diff
@@ -14,50 +14,47 @@ namespace BelissimoCloneWPF.Data.Repositories
...
-            SaveChangesAsync();
+            return (dbContext.Set<T>().Update(entity)).Entity;         
         }
-        public async Task SaveChangesAsync()
+        public async ValueTask SaveChangesAsync()
         {
-            await _dbContext.SaveChangesAsync();
+            await dbContext.SaveChangesAsync();
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:57]: Make it lambda

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 7]
```diff
@@ -14,50 +14,47 @@ namespace BelissimoCloneWPF.Data.Repositories
...
         public GenericRepository(BelissimoDbContext dbContext)
         {
-            _dbContext = dbContext;
+           this.dbContext = dbContext;
         }
         public async ValueTask<T> CreateAsync(T entity)
         {
-            return (await _dbContext.Set<T>().AddAsync(entity)).Entity;
-            SaveChangesAsync();
+            return (await dbContext.Set<T>().AddAsync(entity)).Entity;
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:24]: lmbda

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 8]
```diff
@@ -14,50 +14,47 @@ namespace BelissimoCloneWPF.Data.Repositories
...
-                SaveChangesAsync();
+                dbContext.Set<T>().Remove(entity);             
                 return true;
             }
         }
 
         public IQueryable<T> GetAll(Expression<Func<T, bool>> expression, string[] includes = null, bool isTracking = true)
         {
-            return _dbContext.Set<T>().Where(expression);
+            return dbContext.Set<T>().Where(expression);
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:43]: use include
and isTracking

**Review by MuhammadKasimov — CHANGES_REQUESTED** [review]

**Comment by Dilshod-Abdijabbarov** [comment]
Create lambda expression methods

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 9]
```diff
@@ -14,47 +14,50 @@ namespace BelissimoCloneWPF.Data.Repositories
 {
     public class GenericRepository<T> : IGenericRepository<T> where T : Auditable
     {
-        private readonly BelissimoDbContext dbContext;     
-        public GenericRepository(BelissimoDbContext dbContext)
+        private BelissimoDbContext dbContext; 
+        protected readonly DbSet<T> dbSet;
+        public GenericRepository(BelissimoDbContext dbcontext)
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:19]: not dbcontext
it should be dbContext

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 10]
```diff
@@ -14,47 +14,50 @@ namespace BelissimoCloneWPF.Data.Repositories
...
             if(entity==null)
-            {
-                return false;
-            }
-            else
-            {
-                dbContext.Set<T>().Remove(entity);             
-                return true;
-            }
+               return false;
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:32]: where is if?

File: BelissimoCloneWPF.Data/Repositories/GenericRepository.cs [hunk 11]
```diff
@@ -14,47 +14,50 @@ namespace BelissimoCloneWPF.Data.Repositories
...
+            return true;
         }
 
         public IQueryable<T> GetAll(Expression<Func<T, bool>> expression, string[] includes = null, bool isTracking = true)
         {
-            return dbContext.Set<T>().Where(expression);
+           IQueryable<T> query = expression is null ? dbSet : dbSet.Where(expression);
+
+            if (includes != null)
+                foreach (var includ in includes)
```
Comment (MuhammadKasimov) [inline BelissimoCloneWPF.Data/Repositories/GenericRepository.cs:43]: include* 
not includ

**Review by MuhammadKasimov — CHANGES_REQUESTED** [review]

**Comment by Dilshod-Abdijabbarov** [comment]
Changed GetAll and DeleteAsync methods

**Pull request merged**